src/common.c

```diff
@@ -174,36 +174,53 @@ char*
 str_replace(const char* string, const char* substr,
             const char* replacement)
 {
-    char* tok = NULL;
+    const char* head = NULL;
+    const char* tok = NULL;
     char* newstr = NULL;
-    char* head = NULL;
+    char* wr = NULL;
+    size_t num_substr = 0;
+    size_t len_string, len_substr, len_replacement, len_string_remains, len_newstr;
 
     if (string == NULL)
         return NULL;
 
     if (substr == NULL || replacement == NULL || (strcmp(substr, "") == 0))
         return strdup(string);
 
-    newstr = strdup(string);
-    head = newstr;
+    tok = string;
+    while ((tok = strstr(tok, substr))) {
+        num_substr++;
+        tok++;
+    }
 
-    while ((tok = strstr(head, substr))) {
-        auto_char char* oldstr = newstr;
-        newstr = malloc(strlen(oldstr) - strlen(substr) + strlen(replacement) + 1);
+    if (num_substr == 0)
+        return strdup(string);
 
-        if (newstr == NULL) {
-            return NULL;
-        }
+    len_string = strlen(string);
+    len_substr = strlen(substr);
+    len_replacement = strlen(replacement);
+    len_newstr = len_string - (num_substr * len_substr) + (num_substr * len_replacement);
+    newstr = malloc(len_newstr + 1);
+    if (newstr == NULL) {
+        return NULL;
```

Comment:
I am not sure that it's a very good handling of malloc, since it will make "silent" malfunction. We can discuss it more in MUC.